provider/datagen/src/uprops/casemapping.rs

```diff
@@ -3,24 +3,32 @@
 // (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
 
 use crate::uprops::uprops_serde;
+use crate::DatagenOptions;
 use icu_casemapping::provider::{CaseMappingV1, CaseMappingV1Marker};
 use icu_casemapping::CaseMappingInternals;
 use icu_codepointtrie::CodePointTrieHeader;
 use icu_provider::prelude::*;
 
 use std::convert::TryFrom;
 use std::fs;
-use std::path::PathBuf;
 
 pub struct CaseMappingDataProvider {
     case_mapping: CaseMappingInternals<'static>,
 }
 
 /// A data provider reading from .toml files produced by the ICU4C icuwriteuprops tool.
-impl CaseMappingDataProvider {
-    pub fn try_new(path: PathBuf) -> Result<Self, DataError> {
+impl TryFrom<&DatagenOptions> for CaseMappingDataProvider {
+    type Error = DataError;
+    fn try_from(options: &DatagenOptions) -> Result<Self, Self::Error> {
         #[allow(clippy::unwrap_used)] // TODO(#1668) Clippy exceptions need docs or fixing.
-        let toml_str = fs::read_to_string(&path).unwrap();
+        let toml_str = fs::read_to_string(
+            options
+                .uprops_root
+                .as_ref()
+                .ok_or_else(|| DataError::custom("CaseMappingDataProvider requires uprops root"))?
+                .join("ucase.toml"),
+        )
+        .unwrap();
```

Comment:
Nit: Try to move away from panic in datagen and favor eyre

Reply:
I actually want to move away from `eyre` as well and use `DataError` for when this is called through the future API instead of the binary. I have some code for this already.

Reply:
(so I'd rather remove the panic in a follow-up)